File: problem1_Recursive.cpp
#include <iostream>
#include <vector>

using namespace std;


//Recursive solution

class Solution{
    public:
        int Binary(vector<int>& arr, int target, int start, int end)
        {   
            
            int middle = start + (end - start)/2;
            if(start > end) return -1;
            
            else if(arr[middle] == target) return middle;
            
            else
            {   //left sorted
                if(arr[start] <= arr[middle])
               {
                    if(arr[start] <= target && arr[middle] >= target)
                        return Binary(arr, target, start, middle-1);
                    else    
                       return Binary(arr, target, middle + 1, end);
               }
               
                //right sorted
                else
                {
                    if(arr[middle] <= target && arr[end] >= target)
                        return Binary(arr, target, middle + 1, end);
                    else
                        return Binary(arr, target, start, middle-1);
                }
               
               
            }
            
        }
        
        int Search(vector<int> &arr, int target)
        {
            return Binary(arr, target, 0, arr.size() - 1);
        }
};

int main() {
	//code
	Solution s;
	vector<int>num{4,5,6,7,1,2,3};
	cout<<s.Search(num, 5);
	return 0;
}

File: problem-1-recursive.py
"""
Approach

1. Find the middle element
2. One of the property of roatated sorted array is that either the left part or the right part of the mid is always sorted
3. if the left part is sorted and target in in the range of left part then continue binary search in left part otherwise
search in right part

"""


# Time Complexity : O(log n)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : yes	
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach

class Solution:
    
    def binarySearch(self,nums, left, right,target):    
        if left>right:
            return -1
        mid = left + (right-left)//2
        if nums[mid] == target:
            return mid
        if (nums[left]<=nums[mid]):
            # strictly increaseing
            if target>= nums[left] and target<nums[mid]:
                right = mid-1
            else:
                left=mid+1
        else:                
            if target> nums[mid] and target<=nums[right]:
                left = mid+1
            else:
                right=mid-1
        return self.binarySearch(nums,left,right,target)          
    
          
    def search(self, nums: List[int], target: int) -> int:
        return self.binarySearch(nums,0,len(nums)-1,target)
                        
        